cpp/src/join/sort_merge_join.cu

```diff
@@ -0,0 +1,558 @@
+                          larger_indices.begin(),
+                          mapping_functor<device_span<size_type>>{left_mapping});
+      }
+    }
+    if (is_right_nullable) {
+      auto right_mapping = preprocessed_right.map_table_to_unprocessed(stream);
+      if (is_left_smaller) {
+        thrust::transform(rmm::exec_policy_nosync(stream),
+                          larger_indices.begin(),
+                          larger_indices.end(),
+                          larger_indices.begin(),
+                          mapping_functor<device_span<size_type>>{right_mapping});
+      } else {
+        thrust::transform(rmm::exec_policy_nosync(stream),
+                          smaller_indices.begin(),
+                          smaller_indices.end(),
+                          smaller_indices.begin(),
+                          mapping_functor<device_span<size_type>>{right_mapping});
+      }
+    }
+  }
+}
+
+std::pair<std::unique_ptr<rmm::device_uvector<size_type>>,
+          std::unique_ptr<rmm::device_uvector<size_type>>>
+sort_merge_join::inner_join(table_view const& left,
+                            sorted is_left_sorted,
+                            rmm::cuda_stream_view stream,
+                            rmm::device_async_resource_ref mr)
+{
+  // Sanity checks
+  CUDF_EXPECTS(left.num_columns() != 0,
+               "Number of columns in left keys must be non-zero for a join");
+  CUDF_EXPECTS(left.num_columns() == preprocessed_right._null_processed_table_view.num_columns(),
+               "Number of columns must match for a join");
+
+  // Preprocessing the left table
+  preprocessed_left._table_view = left;
+  if (compare_nulls == null_equality::EQUAL) {
+    preprocessed_left._null_processed_table_view = left;
+  } else {
+    // if a table has no nullable column, then there's no preprocessing to be done
+    auto is_left_nullable = nullable(left);
+    if (is_left_nullable) {
+      preprocessed_left.preprocess_unprocessed_table(stream);
+    } else {
+      preprocessed_left._null_processed_table_view = left;
+    }
+  }
+  if (is_left_sorted == cudf::sorted::NO) { preprocessed_left.get_sorted_order(stream); }
+
+  // Identify the smaller and larger tables for merge operation
+  bool is_left_smaller = preprocessed_left._null_processed_table_view.num_rows() <
+                         preprocessed_right._null_processed_table_view.num_rows();
+  auto& smaller = is_left_smaller ? preprocessed_left : preprocessed_right;
+  auto& larger  = is_left_smaller ? preprocessed_right : preprocessed_left;
+
+  // Neither table was pre-sorted
+  if (smaller._null_processed_table_sorted_order.has_value() &&
+      larger._null_processed_table_sorted_order.has_value()) {
+    merge obj(smaller._null_processed_table_view,
+              smaller._null_processed_table_sorted_order.value()->view().begin<size_type>(),
+              smaller._null_processed_table_sorted_order.value()->view().end<size_type>(),
```

Comment:
Typically, we should avoid calling `view()` on the same object multiple times. It is not a trivial operation. Instead, assigned that `view()` into a local variable and reuse it.